File: relaykit/dto/adaptive_reasoning.go
package dto

import (
	"fmt"
	"slices"
	"strings"
)

// AdaptiveReasoningConfig 仅用于网关决策，不发送给主模型。
type AdaptiveReasoningConfig struct {
	Enabled             bool     `json:"enabled"`
	ChannelID           int      `json:"channel_id"`
	Model               string   `json:"model,omitempty"`
	Efforts             []string `json:"efforts,omitempty"`
	MaxReuseGenerations int      `json:"max_reuse_generations,omitempty"`
	TimeoutMS           int      `json:"timeout_ms,omitempty"`
	MaxChars            int      `json:"max_chars,omitempty"`
}

func (c AdaptiveReasoningConfig) WithDefaults() AdaptiveReasoningConfig {
	if c.Model == "" {
		c.Model = "jev-latest"
	}
	if len(c.Efforts) == 0 {
		c.Efforts = []string{"low", "medium", "high", "xhigh"}
	}
	if c.MaxReuseGenerations == 0 {
		c.MaxReuseGenerations = 10
	}
	if c.TimeoutMS == 0 {
		c.TimeoutMS = 1500
	}
	if c.MaxChars == 0 {
		c.MaxChars = 12000
	}
	return c
}

func (c *AdaptiveReasoningConfig) Validate() error {
	if c == nil || !c.Enabled {
		return nil
	}
	if c.ChannelID <= 0 {
		return fmt.Errorf("adaptive_reasoning.channel_id must be positive")
	}
	if len(c.Model) > 200 || c.Model != strings.TrimSpace(c.Model) {
		return fmt.Errorf("invalid adaptive_reasoning.model")
	}
	if !slices.Contains([]int{0, 1, 2, 5, 10}, c.MaxReuseGenerations) {
		return fmt.Errorf("adaptive_reasoning.max_reuse_generations must be 1, 2, 5 or 10")
	}
	if c.TimeoutMS < 0 || c.TimeoutMS > 30000 {
		return fmt.Errorf("adaptive_reasoning.timeout_ms must be between 0 and 30000")
	}
	if c.MaxChars < 0 || c.MaxChars > 60000 {
		return fmt.Errorf("adaptive_reasoning.max_chars must be between 0 and 60000")
	}
	seen := make(map[string]bool)
	for _, effort := range c.Efforts {
		if !slices.Contains([]string{"none", "minimal", "low", "medium", "high", "xhigh", "max"}, effort) || seen[effort] {
			return fmt.Errorf("invalid or duplicate adaptive_reasoning effort: %s", effort)
		}
		seen[effort] = true
	}
	return nil
}

File: web/src/features/channels/lib/adaptive-reasoning.ts
/*
Copyright (C) 2023-2026 QuantumNous

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of the
License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU Affero General Public License for more details.

You should have received a copy of the GNU Affero General Public License
along with this program. If not, see <https://www.gnu.org/licenses/>.

For commercial licensing, please contact support@quantumnous.com
*/
import { z } from 'zod'

const channelTypes = new Set([1, 48, 57, 58, 61, 62])
export function supportsAdaptiveReasoning(channelType: number): boolean {
  return channelTypes.has(channelType)
}

export const adaptiveReasoningSchema = z
  .object({
    enabled: z.boolean(),
    channel_id: z.number().int().min(0),
    model: z.string().trim().max(200),
    efforts: z.array(
      z.enum(['none', 'minimal', 'low', 'medium', 'high', 'xhigh', 'max'])
    ),
    max_reuse_generations: z
      .number()
      .refine((value) => [1, 2, 5, 10].includes(value), 'Invalid value'),
    timeout_ms: z.number().int().min(1).max(30000),
    max_chars: z.number().int().min(1).max(60000),
  })
  .superRefine((value, ctx) => {
    if (!value.enabled) return
    if (value.channel_id <= 0) {
      ctx.addIssue({
        code: 'custom',
        path: ['channel_id'],
        message: 'Select a TypeSafe channel',
      })
    }
    if (value.efforts.length === 0) {
      ctx.addIssue({
        code: 'custom',
        path: ['efforts'],
        message: 'Select at least one reasoning effort',
      })
    }
  })

export type AdaptiveReasoningConfig = z.infer<typeof adaptiveReasoningSchema>
export const ADAPTIVE_REASONING_DEFAULTS: AdaptiveReasoningConfig = {
  enabled: false,
  channel_id: 0,
  model: 'jev-latest',
  efforts: ['low', 'medium', 'high', 'xhigh'],
  max_reuse_generations: 10,
  timeout_ms: 1500,
  max_chars: 12000,
}

export function readAdaptiveReasoning(value: unknown): AdaptiveReasoningConfig {
  if (!value || typeof value !== 'object' || Array.isArray(value)) {
    return { ...ADAPTIVE_REASONING_DEFAULTS }
  }
  const config = {
    ...ADAPTIVE_REASONING_DEFAULTS,
    ...value,
  } as AdaptiveReasoningConfig
  return {
    ...config,
    model: config.model || ADAPTIVE_REASONING_DEFAULTS.model,
    efforts: config.efforts?.length
      ? config.efforts
      : [...ADAPTIVE_REASONING_DEFAULTS.efforts],
    max_reuse_generations: config.max_reuse_generations || 10,
    timeout_ms: config.timeout_ms || 1500,
    max_chars: config.max_chars || 12000,
  }
}
